Clean up error handling and moves the logic to its own cog

naem-less opened the pull request:
Also removed an unused function and changed the user facing error messages to properly reflect what error was encountered Ensured parity between text and app command error messages (commands.MissingAnyRole and app_commands.MissingAnyRole)

## What does this PR do?

## Fill out the following

<!-- Put an x inside [ ] to check it, like so: [x] -->

- [x] If code changes were made then they have been tested.
- [ ] This PR fixes a bug or issue.
- [ ] This PR adds a new feature.
- [ ] This PR is not a code change (i.e, README)

naem-less committed:
Clean up error handling and moves the logic to its own cog
Also removed an unused function and changed the user facing error messages to properly reflect what error was encountered
Ensured parity between text and app command error messages (commands.MissingAnyRole and app_commands.MissingAnyRole)

naem-less committed:
Improve typehints on format_list function

kiki0124 commented on cogs/error_handler.py:
```diff
@@ -0,0 +1,125 @@
+			await self.handle_text_command_check_failure(ctx, e)
+		elif isinstance(e, commands.CommandOnCooldown):
+			error_message = f"This command is on cooldown for another **{e.retry_after:.2f} seconds**!"
+			await ctx.reply(content=error_message, mention_author=False, allowed_mentions=discord.AllowedMentions(roles=False))
```
Wouldn't it be simpler to just use AllowedMentions.none() ?

kiki0124 replied:
<img width="1446" height="325" alt="image" src="https://github.com/user-attachments/assets/12144406-d828-456a-9e60-ebe49dee10e8" />
Definitely a method

naem-less replied:
just tried it and it works as you can see by the commit :))))))
i have no idea why now and not earlier but its probably something to do with vscode

kiki0124 commented:
Other than that, this seems great. I'll test this sometime in the upcoming week *or* after January 14th. Thank you for your contributions!

naem-less committed:
Use discord.AllowedMentions.none()

Sacul0457 commented on cogs/error_handler.py:
```diff
@@ -0,0 +1,125 @@
+		elif isinstance(e, app_commands.CommandOnCooldown):
+			error_message = f"This command is on cooldown for another **{e.retry_after:.2f} seconds**!"
+		
+		await interaction.response.send_message(error_message, ephemeral=True)
```
```suggestion
	@staticmethod
	async def handle_app_command_check_failure(interaction: discord.Interaction, e: app_commands.CheckFailure) -> None:
		"""
		Handles errors raised by a user not meeting the command check decorators for application commands
		"""
		if isinstance(e, app_commands.NoPrivateMessage):
			error_message = "You may not use this command in DMs!"
		elif isinstance(e, app_commands.MissingRole):
			error_message = f"Only <@&{e.missing_role}> can use this command!"
		elif isinstance(e, app_commands.MissingAnyRole):
			missing_roles = [f"<@&{role_id}>" for role_id in e.missing_roles]
			error_message = f"Only {format_list(missing_roles)} can use this command!"
		elif isinstance(e, app_commands.MissingPermissions):
			missing_perms = "`, `".join(e.missing_permissions)
			error_message = f"Only users with these permissions can use this command: `{missing_perms}`!"
		elif isinstance(e, app_commands.BotMissingPermissions):
			missing_bot_perms = "`, `".join(e.missing_permissions)
			error_message = f"I need the following permissions to be able to execute this command: `{missing_bot_perms}`!"
		elif isinstance(e, app_commands.CommandOnCooldown):
			error_message = f"This command is on cooldown for another **{e.retry_after:.2f} seconds**!"
	    else:
	        error_message = f"Only <@&{DEVELOPERS_ROLE_ID}>, <@&{MODERATORS_ROLE_ID}>, <@&{EXPERTS_ROLE_ID}> and the OP can use this command and only in #support!"
		
		await interaction.response.send_message(error_message, ephemeral=True)
```
[This predicate](https://github.com/kiki0124/Sapphire-helper/blob/69be05382128efc8312fa4ae5e916933078fc32c/cogs/utility.py#L229) raises a generic `app_commands.CheckFailure` I believe, so we need to handle that.

naem-less committed:
Improve format_list typehints to accept more than just lists
Properly handle @app_commands.check() as it can raise app_commands.CheckFailure

Sacul0457 reviewed (approved):
LGTM

kiki0124 committed:
Merge branch 'main' into pr/naem-less/37

kiki0124 reviewed (approved):
Thank you!
